File: binding/src/charts/libqbarcategoryaxis.cpp
#include <QAbstractAxis>
#include <QBarCategoryAxis>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qbarcategoryaxis.h>
#include "libqbarcategoryaxis.h"
#include "libqbarcategoryaxis.hxx"

QBarCategoryAxis* QBarCategoryAxis_new() {
	 return new VirtualQBarCategoryAxis();
}

QBarCategoryAxis* QBarCategoryAxis_new2(QObject* parent) {
	 return new VirtualQBarCategoryAxis(parent);
}

libqt_string QBarCategoryAxis_Tr(const char* s) {
	QString _ret = QBarCategoryAxis::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QBarCategoryAxis_Type(const QBarCategoryAxis* self) {
	return self->type();
}

void QBarCategoryAxis_Append(QBarCategoryAxis* self, const libqt_list categories) {
	self->append(*categories);
}

void QBarCategoryAxis_Append2(QBarCategoryAxis* self, const libqt_string category) {
	self->append(QString::fromUtf8(category.data, category.len));
}

void QBarCategoryAxis_Remove(QBarCategoryAxis* self, const libqt_string category) {
	self->remove(QString::fromUtf8(category.data, category.len));
}

void QBarCategoryAxis_Insert(QBarCategoryAxis* self, int index, const libqt_string category) {
	self->insert(index, QString::fromUtf8(category.data, category.len));
}

void QBarCategoryAxis_Replace(QBarCategoryAxis* self, const libqt_string oldCategory, const libqt_string newCategory) {
	self->replace(QString::fromUtf8(oldCategory.data, oldCategory.len), QString::fromUtf8(newCategory.data, newCategory.len));
}

void QBarCategoryAxis_Clear(QBarCategoryAxis* self) {
	self->clear();
}

void QBarCategoryAxis_SetCategories(QBarCategoryAxis* self, const libqt_list categories) {
	self->setCategories(*categories);
}

libqt_list QBarCategoryAxis_Categories(QBarCategoryAxis* self) {
	return self->categories();
}

int QBarCategoryAxis_Count(const QBarCategoryAxis* self) {
	return self->count();
}

libqt_string QBarCategoryAxis_At(const QBarCategoryAxis* self, int index) {
	QString _ret = self->at(index);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QBarCategoryAxis_SetMin(QBarCategoryAxis* self, const libqt_string minCategory) {
	self->setMin(QString::fromUtf8(minCategory.data, minCategory.len));
}

libqt_string QBarCategoryAxis_Min(const QBarCategoryAxis* self) {
	QString _ret = self->min();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QBarCategoryAxis_SetMax(QBarCategoryAxis* self, const libqt_string maxCategory) {
	self->setMax(QString::fromUtf8(maxCategory.data, maxCategory.len));
}

libqt_string QBarCategoryAxis_Max(const QBarCategoryAxis* self) {
	QString _ret = self->max();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QBarCategoryAxis_SetRange(QBarCategoryAxis* self, const libqt_string minCategory, const libqt_string maxCategory) {
	self->setRange(QString::fromUtf8(minCategory.data, minCategory.len), QString::fromUtf8(maxCategory.data, maxCategory.len));
}

void QBarCategoryAxis_CategoriesChanged(QBarCategoryAxis* self) {
	self->categoriesChanged();
}

void QBarCategoryAxis_Connect_CategoriesChanged(QBarCategoryAxis* self, intptr_t slot) {
    void (*slotFunc)(QBarCategoryAxis*) = reinterpret_cast<void (*)(QBarCategoryAxis*)>(slot);
    QBarCategoryAxis::connect(self, &QBarCategoryAxis::categoriesChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QBarCategoryAxis_MinChanged(QBarCategoryAxis* self, const libqt_string min) {
	self->minChanged(QString::fromUtf8(min.data, min.len));
}

void QBarCategoryAxis_Connect_MinChanged(QBarCategoryAxis* self, intptr_t slot) {
    void (*slotFunc)(QBarCategoryAxis*, const char*) = reinterpret_cast<void (*)(QBarCategoryAxis*, const char*)>(slot);
    QBarCategoryAxis::connect(self, &QBarCategoryAxis::minChanged, [self, slotFunc](const QString& min) {
	slotFunc(self, min);
    });
}

void QBarCategoryAxis_MaxChanged(QBarCategoryAxis* self, const libqt_string max) {
	self->maxChanged(QString::fromUtf8(max.data, max.len));
}

void QBarCategoryAxis_Connect_MaxChanged(QBarCategoryAxis* self, intptr_t slot) {
    void (*slotFunc)(QBarCategoryAxis*, const char*) = reinterpret_cast<void (*)(QBarCategoryAxis*, const char*)>(slot);
    QBarCategoryAxis::connect(self, &QBarCategoryAxis::maxChanged, [self, slotFunc](const QString& max) {
	slotFunc(self, max);
    });
}

void QBarCategoryAxis_RangeChanged(QBarCategoryAxis* self, const libqt_string min, const libqt_string max) {
	self->rangeChanged(QString::fromUtf8(min.data, min.len), QString::fromUtf8(max.data, max.len));
}

void QBarCategoryAxis_Connect_RangeChanged(QBarCategoryAxis* self, intptr_t slot) {
    void (*slotFunc)(QBarCategoryAxis*, const char*, const char*) = reinterpret_cast<void (*)(QBarCategoryAxis*, const char*, const char*)>(slot);
    QBarCategoryAxis::connect(self, &QBarCategoryAxis::rangeChanged, [self, slotFunc](const QString& min, const QString& max) {
	slotFunc(self, min, max);
    });
}

void QBarCategoryAxis_CountChanged(QBarCategoryAxis* self) {
	self->countChanged();
}

void QBarCategoryAxis_Connect_CountChanged(QBarCategoryAxis* self, intptr_t slot) {
    void (*slotFunc)(QBarCategoryAxis*) = reinterpret_cast<void (*)(QBarCategoryAxis*)>(slot);
    QBarCategoryAxis::connect(self, &QBarCategoryAxis::countChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

libqt_string QBarCategoryAxis_Tr2(const char* s, const char* c) {
	QString _ret = QBarCategoryAxis::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QBarCategoryAxis_Tr3(const char* s, const char* c, int n) {
	QString _ret = QBarCategoryAxis::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
int QBarCategoryAxis_QBaseType(const QBarCategoryAxis* self) {
	auto* vqbarcategoryaxis = dynamic_cast<const VirtualQBarCategoryAxis*>(self);
	if (vqbarcategoryaxis && vqbarcategoryaxis->isVirtualQBarCategoryAxis) {
vqbarcategoryaxis->setQBarCategoryAxis_Type_IsBase(true);
	return vqbarcategoryaxis->type();
}
}

// Auxiliary method to allow providing re-implementation
void QBarCategoryAxis_OnType(const QBarCategoryAxis* self, intptr_t slot) {
	auto* vqbarcategoryaxis = dynamic_cast<const VirtualQBarCategoryAxis*>(self);
	if (vqbarcategoryaxis && vqbarcategoryaxis->isVirtualQBarCategoryAxis) {
vqbarcategoryaxis->setQBarCategoryAxis_Type_Callback(reinterpret_cast<VirtualQBarCategoryAxis::QBarCategoryAxis_Type_Callback>(slot));
}
}

void QBarCategoryAxis_Delete(QBarCategoryAxis* self) {
    delete self;
}

File: binding/src/charts/libqbarcategoryaxis.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qbarcategoryaxis.html
pub const qbarcategoryaxis = struct {

    /// New constructs a new QBarCategoryAxis object.
    pub fn New() QtC.QBarCategoryAxis {
        return qtc.QBarCategoryAxis_new();
    }


    /// New2 constructs a new QBarCategoryAxis object.
    pub fn New2(parent: ?*anyopaque) QtC.QBarCategoryAxis {
        return qtc.QBarCategoryAxis_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Type(self: ?*anyopaque, ) i32 {
        return qtc.QBarCategoryAxis_Type(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnType(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QBarCategoryAxis_OnType(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseType(self: ?*anyopaque, ) i32 {
        return qtc.QBarCategoryAxis_QBaseType(@ptrCast(self));
    }

    pub fn Append(self: ?*anyopaque, categories: []const u8) void {
        qtc.QBarCategoryAxis_Append(@ptrCast(self), @ptrCast(categories));
    }

    pub fn Append2(self: ?*anyopaque, category: []const u8) void {
        const category_str = qtc.libqt_string{
    .len = category.len,
    .data = category.ptr,
};
qtc.QBarCategoryAxis_Append2(@ptrCast(self), category_str);
    }

    pub fn Remove(self: ?*anyopaque, category: []const u8) void {
        const category_str = qtc.libqt_string{
    .len = category.len,
    .data = category.ptr,
};
qtc.QBarCategoryAxis_Remove(@ptrCast(self), category_str);
    }

    pub fn Insert(self: ?*anyopaque, index: i32, category: []const u8) void {
        const category_str = qtc.libqt_string{
    .len = category.len,
    .data = category.ptr,
};
qtc.QBarCategoryAxis_Insert(@ptrCast(self), index, category_str);
    }

    pub fn Replace(self: ?*anyopaque, oldCategory: []const u8, newCategory: []const u8) void {
        const oldCategory_str = qtc.libqt_string{
    .len = oldCategory.len,
    .data = oldCategory.ptr,
};
const newCategory_str = qtc.libqt_string{
    .len = newCategory.len,
    .data = newCategory.ptr,
};
qtc.QBarCategoryAxis_Replace(@ptrCast(self), oldCategory_str, newCategory_str);
    }

    pub fn Clear(self: ?*anyopaque, ) void {
        qtc.QBarCategoryAxis_Clear(@ptrCast(self));
    }

    pub fn SetCategories(self: ?*anyopaque, categories: []const u8) void {
        qtc.QBarCategoryAxis_SetCategories(@ptrCast(self), @ptrCast(categories));
    }

    pub fn Categories(self: ?*anyopaque, ) []const u8 {
        return qtc.QBarCategoryAxis_Categories(@ptrCast(self));
    }

    pub fn Count(self: ?*anyopaque, ) i32 {
        return qtc.QBarCategoryAxis_Count(@ptrCast(self));
    }

    pub fn At(self: ?*anyopaque, index: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_At(@ptrCast(self), index);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.At: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetMin(self: ?*anyopaque, minCategory: []const u8) void {
        const minCategory_str = qtc.libqt_string{
    .len = minCategory.len,
    .data = minCategory.ptr,
};
qtc.QBarCategoryAxis_SetMin(@ptrCast(self), minCategory_str);
    }

    pub fn Min(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_Min(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.Min: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetMax(self: ?*anyopaque, maxCategory: []const u8) void {
        const maxCategory_str = qtc.libqt_string{
    .len = maxCategory.len,
    .data = maxCategory.ptr,
};
qtc.QBarCategoryAxis_SetMax(@ptrCast(self), maxCategory_str);
    }

    pub fn Max(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_Max(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.Max: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetRange(self: ?*anyopaque, minCategory: []const u8, maxCategory: []const u8) void {
        const minCategory_str = qtc.libqt_string{
    .len = minCategory.len,
    .data = minCategory.ptr,
};
const maxCategory_str = qtc.libqt_string{
    .len = maxCategory.len,
    .data = maxCategory.ptr,
};
qtc.QBarCategoryAxis_SetRange(@ptrCast(self), minCategory_str, maxCategory_str);
    }

    pub fn CategoriesChanged(self: ?*anyopaque, ) void {
        qtc.QBarCategoryAxis_CategoriesChanged(@ptrCast(self));
    }

    pub fn OnCategoriesChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QBarCategoryAxis_Connect_CategoriesChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MinChanged(self: ?*anyopaque, min: []const u8) void {
        const min_str = qtc.libqt_string{
    .len = min.len,
    .data = min.ptr,
};
qtc.QBarCategoryAxis_MinChanged(@ptrCast(self), min_str);
    }

    pub fn OnMinChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QBarCategoryAxis_Connect_MinChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn MaxChanged(self: ?*anyopaque, max: []const u8) void {
        const max_str = qtc.libqt_string{
    .len = max.len,
    .data = max.ptr,
};
qtc.QBarCategoryAxis_MaxChanged(@ptrCast(self), max_str);
    }

    pub fn OnMaxChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8) callconv(.c) void) void {
        qtc.QBarCategoryAxis_Connect_MaxChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn RangeChanged(self: ?*anyopaque, min: []const u8, max: []const u8) void {
        const min_str = qtc.libqt_string{
    .len = min.len,
    .data = min.ptr,
};
const max_str = qtc.libqt_string{
    .len = max.len,
    .data = max.ptr,
};
qtc.QBarCategoryAxis_RangeChanged(@ptrCast(self), min_str, max_str);
    }

    pub fn OnRangeChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, [*:0]const u8, [*:0]const u8) callconv(.c) void) void {
        qtc.QBarCategoryAxis_Connect_RangeChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn CountChanged(self: ?*anyopaque, ) void {
        qtc.QBarCategoryAxis_CountChanged(@ptrCast(self));
    }

    pub fn OnCountChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QBarCategoryAxis_Connect_CountChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QBarCategoryAxis_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qbarcategoryaxis.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QBarCategoryAxis_Delete(@ptrCast(self));
    }
};
